src/lib/ledger_builder/ledger_builder.ml

```diff
@@ -301,81 +296,134 @@ end = struct
+            let open Deferred.Or_error.Let_syntax in
+            let%bind merged_statement =
+              Deferred.return
+              @@ Ledger_proof_statement.merge
+                   (Ledger_proof.statement proof_1)
+                   (Ledger_proof.statement proof_2)
+            in
+            merge_acc
+              ~verify_proof:(fun () ->
+                Deferred.List.for_all
+                  [(proof_1, message_1); (proof_2, message_2)] ~f:
+                  (fun (proof, message) ->
+                    Inputs.Ledger_proof_verifier.verify ~message proof
+                      (Ledger_proof.statement proof) ) )
+              acc_statement merged_statement
+        | Base None -> Deferred.Or_error.return acc_statement
+        | Base
+            (Some
+              { Super_transaction_with_witness.transaction_with_info
+              ; statement
+              ; witness }) ->
+            with_error "Bad base statement" ~f:(fun () ->
+                let expected_statement =
+                  let open Or_error.Let_syntax in
+                  let source =
+                    Frozen_ledger_hash.of_ledger_hash
+                    @@ Sparse_ledger.merkle_root witness
+                  in
+                  let%bind transaction =
+                    Ledger.Undo.super_transaction transaction_with_info
+                  in
+                  let%bind after =
+                    Or_error.try_with (fun () ->
+                        Sparse_ledger.apply_super_transaction_exn witness
+                          transaction )
+                  in
+                  let target =
+                    Frozen_ledger_hash.of_ledger_hash
+                    @@ Sparse_ledger.merkle_root after
+                  in
+                  let%bind fee_excess =
+                    Super_transaction.fee_excess transaction
+                  in
+                  let%map supply_increase =
+                    Super_transaction.supply_increase transaction
+                  in
+                  { Ledger_proof_statement.source
+                  ; target
+                  ; fee_excess
+                  ; supply_increase
+                  ; proof_type= `Base }
+                in
+                let open Deferred.Or_error.Let_syntax in
+                let%bind expected_statement =
+                  Deferred.return expected_statement
+                in
+                if Ledger_proof_statement.equal statement expected_statement
+                then
+                  merge_acc
+                    ~verify_proof:(fun () -> Deferred.return true)
+                    acc_statement statement
+                else
+                  Deferred.Or_error.error_string
+                    (write_error "Bad base statement") )
+      in
+      let res =
+        let open Container.Continue_or_stop in
+        Parallel_scan.State.Deferred.fold_chronological_until t ~init:(Ok None)
```

Comment:
I had the accumulator be `Or_error.t` because I wanted to propagate the `Error.t` from `fold_step`. My fold_chronological_until forces the user to return one type, thus I had to return `Or_error.t` to propagate the error. I will just change the implementation of `fold_chronological_until` to return the union type ('a, 'b) Continue_or_Stop.t, so that this step returns  `Continue Option.t | Stop Error.t`. This would lead to less pattern matching and the code will be simpler to understand.